### PR TITLE
Fix up Command Debug output when arg0 is specified.

### PR DESCRIPTION
**Pull request by jsgf**
PR https://github.com/rust-lang/rust/pull/66512 added the ability to set argv[0] on
Command. As a side effect, it changed the Debug output to print both the program and
argv[0], which in practice results in stuttery output (`"echo" "echo" "foo"`).

This PR reverts the behaviour to the the old one, so that the command is only printed
once - unless arg0 has been set. In that case it emits `"[command]" "arg0" "arg1" ...`.

### CODE REVIEW

File: src/test/ui/command-argv0-debug.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// run-pass
```
Comment (Centril) [inline src/test/ui/command-argv0-debug.rs]: Please move this test to some sub-folder. Dumping things right in `ui/` clutters the overall organization of tests.

Comment (jsgf) [reply src/test/ui/command-argv0-debug.rs]: OK - it's adjacent to the existing tests for command. Should I move them all?

Comment (Centril) [reply src/test/ui/command-argv0-debug.rs]: That would be great, thanks!

Comment (jsgf) [reply src/test/ui/command-argv0-debug.rs]: OK, done.

File: src/libstd/sys/unix/process/process_common.rs [hunk 2]
```diff
@@ -375,8 +375,12 @@ impl ChildStdio {
 
 impl fmt::Debug for Command {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "{:?}", self.program)?;
-        for arg in &self.args {
+        if self.program == self.args[0] {
+            write!(f, "{:?}", self.program)?;
+        } else {
+            write!(f, "[{:?}] {:?}", self.program, self.args[0])?;
+        }
```
Comment (joshtriplett) [inline src/libstd/sys/unix/process/process_common.rs]: As a minor nit, could you common up the branches of this and avoid the else, like this:

```rust
if self.program != self.args[0] {
    write!(f, "[{:?}] ", self.program)?;
}
write!(f, "{:?}", self.args[0])?;
```

(It probably isn't worth trying to join this with the following loop and get the spaces right.)

Comment (jsgf) [reply src/libstd/sys/unix/process/process_common.rs]: Done

**Comment by Mark-Simulacrum** [comment]
I am nominating this for beta backport as it seems it's regressing some crates (i.e., this PR landed too late to fix beta as well). I tagged with T-compiler and T-libs; but I feel fairly confident that we just want to backport this. If T-compiler decides not to approve we'll want to ping T-libs as they don't generally notice otherwise I think.


* root: Srynetix/pssh-rs: [start](https://crater-reports.s3.amazonaws.com/beta-1.41-1/1.40.0/gh/Srynetix.pssh-rs/log.txt) v. [end](https://crater-reports.s3.amazonaws.com/beta-1.41-1/beta-2019-12-18/gh/Srynetix.pssh-rs/log.txt)
 * root: submodules-0.1.0: [start](https://crater-reports.s3.amazonaws.com/beta-1.41-1/1.40.0/reg/submodules-0.1.0/log.txt) v. [end](https://crater-reports.s3.amazonaws.com/beta-1.41-1/beta-2019-12-18/reg/submodules-0.1.0/log.txt)

**Comment by jsgf** [comment]
@Mark-Simulacrum Do any of them seem to be using the Debug output for real functionality, or is it just unit tests failing?

**Comment by Mark-Simulacrum** [comment]
AFAICT, just unit tests - though some are calling public functions, so a little unclear. The duplication of the executable though is somewhat annoying (and will affect other places, e.g. rustc output on failure to run linkers, I imagine).

**Comment by Mark-Simulacrum** [comment]
@rust-lang/libs -- this is nominated for beta backport, and T-compiler (likely correctly :) decided that they feel that libs team should approve this. @pnkfelix will also write up a comment hopefully with some thoughts in an hour or so.

**Comment by pnkfelix** [comment]
Oh yeah, I promised a note here.

My instinct is that I wouldn't backport this PR, since it is inserting this new square-bracket delimted `[command]`. To me, a change like that seems like it would be better off riding the trains.

At the same time, there is a very real regression that should be addressed in some manner.

So what I would advise (and I'm speaking with basically zero authority here, since I'm not really designed with any libs design, apart from at best the `Allocator` API and the defunct placement-in APIs) is to backport a PR reverting *all* of #66512 from beta. But keep #66512 on nightly, and let this PR land there.

----

Before I started looking at this bug, I would have said that I believe our intent is that we are free to change `std::fmt::Debug` implementations for stdlib types at will. However, I am not sure if that policy is set in stone anywhere, so I am starting to doubt my own understanding of the guarantees here. The docs for the [`Debug` trait](https://doc.rust-lang.org/std/fmt/trait.Debug.html) and for [`std::fmt`](https://doc.rust-lang.org/std/fmt/index.html) do not seem to make any strong statements giving stdlib the freedom that I thought it had to revise `Debug` implementations at will.

**Comment by sfackler** [comment]
Debug representations are subject to change at any time.

**Comment by Mark-Simulacrum** [comment]
To be clear, I consider the behavior on beta at least suboptimal (double printing the program), and plausibly a bug. Given instability of Debug, though, we may decide to not actually backport anything and merely note it in the release notes.

**Comment by jsgf** [comment]
The `["executable"]` output will only appear if you're using the new API. With this change, the Debug output is unchanged for normal uses and also avoid the double-printing. Without this change the Debug output changes for no reason and in a generally unhelpful way.

I see this as being a bugfix to the original PR and I think it would be best to backport it so that they both land together. Otherwise we'll end up with a release with a Debug output regression.

**Comment by alexcrichton** [comment]
I personally think it's fine to backport this, so I'm going to accept it.

**Comment by jsgf** [comment]
@alexcrichton What's the actual process for backporting? Should I submit a PR?

**Comment by Mark-Simulacrum** [comment]
I will be doing a beta backport rollup today or tomorrow probably and this'll get picked up since it's been approved, so no need to do anything further.

**Pull request merged**